storage/api.go

```diff
@@ -0,0 +1,95 @@
+// which may cause some performance overhead when handling many small writes in sequence.
+//
+// The PutStream function returns three parameters: an io.Writer (as you'd expect), another function, and an error.
+// The function returned is called a "WriteCommitter".
+// The final error value is as usual: it will contain an error value if the write could not be begun.
+// ("WriteCommitter" will be refered to as such throughout the docs, but we don't give it a named type --
+// unfortunately, this is important, because we don't want to force implementers of storage systems to import this package just for a type name.)
+//
+// The WriteCommitter function should be called when you're done writing,
+// at which time you give it the key you want to commit the data as.
+// It will close and flush any streams, and commit the data to its final location under this key.
+// (If the io.Writer is also an io.WriteCloser, it is not necessary to call Close on it,
+// because using the WriteCommiter will do this for you.)
+//
+// Because these storage APIs are meant to work well for content-addressed systems,
+// the key argument is not provided at the start of the write -- it's provided at the end.
+// (This gives the opportunity to be computing a hash of the contents as they're written to the stream.)
+//
+// As a special case, giving a key of the zero string to the WriteCommiter will
+// instead close and remove any temp files, and store nothing.
+// An error may still be returned from the WriteCommitter if there is an error cleaning up
+// any temporary storage buffers that were created.
+//
+// Continuing to write to the io.Writer after calling the WriteCommitter function will result in errors.
+// Calling the WriteCommitter function more than once will result in errors.
+type StreamingWritableStorage interface {
+	PutStream(ctx context.Context) (io.Writer, func(key string) error, error)
+}
+
+// --- other specializations --->
+
+// VectorWritableStorage is an API for writing several slices of bytes at once into storage.
+// It's meant a feature-detection interface; not all storage implementations need to provide this feature.
+// This kind of API can be useful for maximizing performance in scenarios where
+// data is already loaded completely into memory, but scattered across several non-contiguous regions.
+type VectorWritableStorage interface {
+	PutVec(ctx context.Context, key string, blobVec [][]byte) error
+}
+
+// PeekableStorage is a feature-detection interface which a storage implementation can use to advertise
+// the ability to look at a piece of data, and return it in shared memory.
+// The PeekableStorage.Peek method is essentially the same as ReadableStorage.Get --
+// but by contrast, ReadableStorage is expected to return a safe copy.
+// PeekableStorage can be used when the caller knows they will not mutate the returned slice.
+type PeekableStorage interface {
+	Peek(ctx context.Context, key string) ([]byte, error)
```

Comment:
Another option, if you want to avoid callbacks, is also returning an `io.Closer` to signal that you're done using the returned bytes: https://pkg.go.dev/github.com/cockroachdb/pebble#Batch.Get

Reply:
Request: please review the interface of candidate backing implementations (like the one @whyrusleeping posted) before choosing a direction here. Getting creative with interfaces, while fun, may lead to poor choices. io.Closer is an interface that I would despise using, as it precludes the calling code from using closures.